### DISCUSSION TITLE
How can I iterate a JavaScript iterator in Python?

### DISCUSSION THREAD

**Discussion by CNSeniorious000** [pyodide/pyodide#4360] [Q&A] [answered]
I need to call a sync iterator in JavaScript (or an async one iterator), is it possible to pass the iterator through the `js.` namespace?

Or, can I call a function from `js.` in python, which will return a js async iterator? Can I iterate it in python using `async for`?

**Comment by hoodmane** [comment, marked as answer]
Yeah this possible. JavaScript (async) iterables are Python (async) iterables. If you define the JS function in global scope then you can import it from `js`, or you can directly assign to the Python global scope with `pyodide.globals.set`. For example:

#### Synchronous:
```js
function* f() {
   yield 1;
   yield 2;
   yield 3;
}
pyodide.globals.set("f", f);
pyodide.runPython(`
for x in f():
   print(x)
`);
```

#### Asynchronous:
```js
function sleep(ms) {
   return new Promise(res => setTimeout(res, ms));
}
async function* f() {
   yield 1;
   await sleep(1000);
   yield 2;
   await sleep(1000);
   yield 3;
}
pyodide.globals.set("f", f);
pyodide.runPythonAsync(`
async for x in f():
   print(x)
`);
```

**Reply by CNSeniorious000** [reply]
Seems great, I will try this later. Thanks a lot!

I am trying to stream openai chat completions response at client side, but I want to use a python-only prompt engineering framework, so I am considering this approach.
